### DISCUSSION TITLE
Filtering and limiting back relations?

### DISCUSSION THREAD

**Discussion by adamshand** [pocketbase/pocketbase#5036] [Q&A] [answered]
Hello again,

The below query works as expected.  Look up user based on URL param and expand `observations` via `tracker` (relation pointing at `users`).  

```
const tracker = await pb.collection('users').getFullList({
	expand: 'observations_via_tracker',
	filter: `username = "${params.tracker}"`
});
```

Is there a way to limit the number of returned observation records?

Is there a way to filter the returned observation records?  I want to only expand records where `observation.visibility = 'public'`.  

It seems like the below should work, but it doesn't.  It doesn't error, but it also doesn't return any results (including the matching user).

```
const tracker = await pb.collection('users').getFullList({
	expand: 'observations_via_tracker',
	filter: `username = "${params.tracker}" && observations_via_tracker.visibility = "public"`,
});
```

Any pointers much appreciated, thanks!

**Comment by ganigeorgiev** [comment]
`filter` and `sort` applies only for the main queried record (in your case `users`).

`expand` expands all records that the user can "view" no matter of the main `filter` value.

By having `observations_via_tracker.visibility = "public"` in your filter constraint this will return only `users` records with `observations_via_tracker.visibility` is public, aka. the `filter` skips/filters `users` records, not `observations_via_tracker` expands.

If you want to apply a filter on `expand` result, you could:

- apply the expanded filter on the client-side
- change the View API rule of the `observations` collection (_in the next release expand checks will have their own `@request.context = "expand"` that will allow applying a conditional constraints only when used with `expand`_)
- instead of `expand`, send a separate request to the related `observations_via_tracker` collection and perform the "users" grouping on the client-side (you can filter observations by the user relation id(s)). An example pseudo-code could be:
    ```js
    const users = await pb.collection('users').getFullList({ // maybe replace with getFirstListItem if always targeting only 1 user?
        filter: pb.filter("username={:username}", { username: params.tracker }),
    });

    // retrieve only observations_via_tracker that are public
    const observations = await pb.collection("observations").getFullList({
        // "visibility="public" &&  (tracker = id1 || tracker = id2 || ...)"
        filter: 'visibility="public" && (' + users.map((id) => `tracker="${id}"`).join("||") + ")",
    })

    // manual group observation per user if needed...
    ```
    (_edit: updated the filter to reflect the "tracker" relation field_)

**Reply by adamshand** [reply]
That's very helpful.  

I'm curious if there is there a reason you used:
```
pb.filter("username={:username}", { username: params.tracker })
```

I think it does the same thing as what I originally had?
```
filter: `username = "${params.tracker}"`
```

I'll hopefully end up with many thousands of observations, so probably makes sense to do a second query.

Thank you!

**Reply by ganigeorgiev** [reply, marked as answer]
`pb.filter` will produce the same but it will also take care for escaping the parameter. This is useful in case it is executed on the server and comes from entrusted/unverified user input. See also [Binding `filter` parameters](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#binding-filter-parameters).

**Reply by adamshand** [reply]
Got it, thanks!

**Comment by jordant97** [comment]
One issue I've encountered in my own code is expanding a field where I only need the last few records. For example, when fetching users, I might want to retrieve just the latest 3 logins for each user. If I have 50 users and initiate 50 separate requests to fetch the last 3 logins for each, it feels quite inefficient.

While this can be addressed using pb_hooks, I've found a helpful alternative solution: including the relation directly in the users collection. By adding a 'lastLogins' field that points to the IDs of the last 3 logins, we can significantly reduce the number of requests and improve efficiency.

I'm just sharing this here in case someone finds it helpful. It's a simple approach that has worked well for me.
